### PR TITLE
Return correct process exit code for browser processes

### PR DESCRIPTION
**Pull request by vritant24**
With a recent change, the `terminate` method in the `BrowserWasiProess` resolved the `runPromise` to ensure that the caller gets a return value on calling `WasmProcess.run()`. 
This had an unintended consequence of always resolving the `runPromise` with `0` as the `proc_exit` method calls `terminate` before resolving the `runPromise` to clean up the resources.

This change separates "terminating" and "cleaning up resources" and makes the `proc_exit` method only clean up resources. This ensures that the `BrowserWasiProcess` returns the correct exit code on the process exiting.

### CODE REVIEW

File: wasm-wasi-core/src/desktop/process.ts [hunk 1]
```diff
@@ -104,14 +104,19 @@ export class NodeWasiProcess extends WasiProcess {
 		return Promise.resolve();
 	}
 
+	protected async cleanupResources(): Promise<void> {
+		await this.mainWorker?.terminate();
```
Comment (vritant24) [inline wasm-wasi-core/src/desktop/process.ts:108]: This is the only part I'm unsure of. I did not find an easy way to both allow the base class to terminate the `mainWorker` to clean the resource, and also allow the `terminate` method to get the result from termination. So for now, I've duplicated the calls assuming calling terminate twice shouldn't be an issue.

File: wasm-wasi-core/src/common/process.ts [hunk 2]
```diff
@@ -296,6 +296,8 @@ export abstract class WasiProcess {
 		});
 	}
 
+	protected abstract processEnded(isExternallyTerminated: boolean): Promise<void>;
```
Comment (dbaeumer) [inline wasm-wasi-core/src/common/process.ts:299]: I would actually call this procExit() and not have a boolean here. We can then in the subclasses decide if we have a helper method to share or if we inline the same code in both method. This makes it IMO easier to understand

**Review by dbaeumer — CHANGES_REQUESTED** [review]

**Review by dbaeumer — CHANGES_REQUESTED** [review]

File: wasm-wasi-core/src/desktop/process.ts [hunk 3]
```diff
@@ -104,14 +107,22 @@ export class NodeWasiProcess extends WasiProcess {
 		return Promise.resolve();
 	}
 
+	protected async procExit(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (vritant24) [reply wasm-wasi-core/src/desktop/process.ts:107]: understood, made it so that the cleanup is inlined in both

**Review by dbaeumer — APPROVED** [review]

**Review by aeschli — APPROVED** [review]

**Comment by vritant24** [comment]
Thanks for the merge!

**Pull request merged**